### PR TITLE
exp(p3): add --gae-lambda CLI + #282 high-λ sweep infrastructure

### LINKED ISSUE #282: exp(p3): high-λ GAE PPO smoke — is bootstrap bias the misalignment source?
## Thesis context

Per `research_notebook/2026-05-17_thesis_misaligned_gradients.md`, the working hypothesis is that PPO's per-step advantage estimator is *misaligned* with the long-horizon cooperative objective. GAE-λ interpolates between high-bias bootstrap (λ=0) and high-variance Monte Carlo (λ=1). Pushing λ → 1 gives more Monte Carlo credit, less bootstrap bias.

**Cheapest possible test of the thesis.** If shifting λ → 0.99 lifts the verdict, our misalignment claim is "GAE bootstrap bias is the source."

## Protocol

- 50 iterations × 3 seeds on `minimal_specialization`
- λ ∈ {0.95 (current default), 0.99, 0.999, 1.0}
- All other hyperparameters held constant; just sweep λ
- Compare gap_closed against the post-#236 baseline (0.182 from PR #257)

## Open question to confirm first

Was high-λ already covered in #174's earlier hyperparameter ablation? If so, lift the result and close as duplicate. If not, run it — it's a 1-line config change + ~30 min wall-clock.

## Verdict logic

| Outcome | Interpretation |
|---|---|
| λ=1.0 reaches gap_closed ≥ 0.5 | Bootstrap bias was the misalignment. PPO needs Monte Carlo. |
| Flat across λ | Bootstrap isn't the dominant misalignment source. Other interventions needed. |
| λ → 1 hurts (variance dominates) | Standard finding; nothing new. |

## Sequencing

Run AFTER #270 verdict lands. If basin trap: still valuable. If anti-attractor: even more valuable (directly tests one anti-attractor mechanism).

## References

- `research_notebook/2026-05-17_thesis_misaligned_gradients.md` (thesis context)
- GAE paper (Schulman et al. 2016)
- PR #257 (#239 verdict; obs-fix baseline)

---

### Hypothesis

GAE-λ interpolates per-step credit assignment between high-bias TD bootstrap (`λ=0`) and high-variance Monte Carlo (`λ=1`). The post-#277 thesis (`research_notebook/2026-05-17_thesis_misaligned_gradients.md`) identifies *gradient locality* — not gradient direction — as the source of PPO's plateau on `minimal_specialization`. Pushing `λ → 1` reduces bootstrap bias by lengthening the effective credit horizon. **If bootstrap bias is the dominant misalignment source, raising λ should lift `gap_closed` from the ~0.18 plateau toward specialist (≥ 0.5).** If λ has little effect, bootstrap bias is not the binding constraint and a deeper intervention (LOLA, hierarchical RL, potential-based shaping) is required.

This is a **smoke test** — short runs, low compute, one-line config sweep. Verdict in ~30 min wall-clock on `COMPUTE_HOST_PRIMARY`.

### Confirm-first question

The #174 plateau ablation (`experiments/p3_specialization/run_174_ablation.sh:19-27`) swept `normalize_returns × value_coef × entropy_coef` only — **GAE-λ was not part of that sweep**. The current code defaults `gae_lambda = 0.95` across the whole stack (`bucket_brigade/training/joint_trainer.py:202`, `bucket_brigade/training/policy_learner.py:47`, `bucket_brigade/training/networks.py:274`). So the experiment is novel; no prior result to lift.

### Protocol

- **Scenario:** `minimal_specialization`
- **λ grid:** `{0.95 (current default — re-run for matched comparison), 0.99, 0.999, 1.0}`
- **Seeds:** 3 (42, 43, 44) — matches #270 / #239 conventions
- **Iterations:** 50 × 2048 rollout steps (matches #270 baseline)
- **Held constant:** `lambda_red=0.0`, IPPO (no `--centralized-critic`), `value_coef=0.5`, `entropy_coef=0.01`, `normalize_returns=False`, no curriculum, no action-shaping, no BC-init
- **Compute:** Remote. Run on `COMPUTE_HOST_PRIMARY` per `CLAUDE.md` compute guidelines (NEVER locally — this is a PPO sweep)

### Baseline comparison

Post-#236 / PR #257 baseline on `minimal_specialization` IPPO + random init: `gap_closed ≈ 0.182` (see `research_notebook/2026-05-17_issue272_bc_fit_only_verdict.md` and `experiments/p3_specialization/analyze_270.py:163`). The λ=0.95 cell in this sweep doubles as a re-baseline; the λ ∈ {0.99, 0.999, 1.0} cells are the experimental arms.

### Success criterion (tier verdict, matching `analyze_261_calibration.py:25-28`)

| Best-cell `gap_closed_mean` | Tier | Interpretation |
|---|---|---|
| ≥ 0.50 | tier_1_breaks_plateau | **Bootstrap bias was the misalignment source.** PPO needs Monte Carlo credit on this game. Promotes #285 (BC-init + high-λ) and motivates production training change. |
| 0.25 - 0.50 | tier_2_partial | Bootstrap is contributing but not sole source. Combine with #285 / other interventions. |
| < 0.25 | tier_3_insufficient | Bootstrap is not the binding constraint. Direct hypothesis pivot to LOLA / potential-based shaping / hierarchical RL. |
| λ → 1 strictly hurts | (expected, prior) | Standard variance-dominance finding; nothing new about misalignment. |

### Implementation guidance

**Required code change** (`experiments/p3_specialization/train.py`): The current CLI does not expose `--gae-lambda`. Add it:

- Add `gae_lambda: float = 0.95` to `CellConfig` (around `train.py:60-95`).
- Add `p.add_argument("--gae-lambda", type=float, default=0.95)` to the argparse block (around `train.py:767-864`).
- Thread `gae_lambda=cfg.gae_lambda` into the `JointPPOTrainer(...)` constructor call (the trainer already accepts it at `bucket_brigade/training/joint_trainer.py:202`, so this is purely a wiring change).
- Log the resolved λ value into `config.json` and `metrics.json` so the analyzer can group runs by λ.

**Sweep driver** (new file, mirror `run_issue270_ppo_continuation.sh`):

```bash
# experiments/p3_specialization/run_issue282_lambda_sweep.sh
# 4 λ cells × 3 seeds = 12 runs
for LAMBDA in 0.95 0.99 0.999 1.0; do
  for SEED in 42 43 44; do
    OUT=experiments/p3_specialization/runs/issue282_lambda_sweep/minimal_specialization/lambda_${LAMBDA}/seed_${SEED}
    mkdir -p "$OUT"
    python experiments/p3_specialization/train.py \
      --scenario minimal_specialization \
      --lambda-red 0.0 \
      --seed "$SEED" \
      --gae-lambda "$LAMBDA" \
      --num-iterations 50 \
      --rollout-steps 2048 \
      --num-agents 4 \
      --output-dir "$OUT" 2>&1 | tee "$OUT/train.log"
  done
done
```

**Analyzer** (new file, model on `analyze_270.py`):
- Load each cell's `metrics.json`, compute trailing-5-iteration mean team reward, run through `gap_closed(...)` from `analyze_270.py:46-58` (uses the canonical `minimal_specialization` random/specialist references).
- Emit a per-λ table (`gap_closed_mean`, `gap_closed_per_seed`, `team_reward_trailing5`) and a tier verdict per cell.

### Sequencing

Run AFTER PR #278 verdict (issue #270 specialist-BC + IPPO continuation) lands — that verdict frames whether the cooperative basin is *stable* under PPO updates at all. Independent of #285 (BC-init + high-λ); #282 is the prerequisite smoke test that decides whether λ matters before doing the combined experiment.

### Relationship to other open issues

- **#285** (BC-init + high-λ combined): *Related but distinct.* This issue (#282) is the standalone λ sweep from random init — answers "does λ matter at all?" before #285 spends effort on the combined experiment. Not a duplicate; #285 explicitly references "sister issue" #282.
- **#270** (closed/building, PR #278): provides the BC-init + IPPO continuation infrastructure #285 builds on. Independent of #282.
- **#174** (closed): the prior ablation; did NOT cover GAE-λ. No overlap.
- **#189**: file/issue search returned no GAE config issue at this number in current state; the existing GAE default of 0.95 has been in place since `JointPPOTrainer` was authored.

## Affected Files

- `experiments/p3_specialization/train.py` — Add `gae_lambda` to `CellConfig`, add `--gae-lambda` CLI arg, thread into `JointPPOTrainer(...)` constructor call.
- `experiments/p3_specialization/run_issue282_lambda_sweep.sh` (new) — 4 λ × 3 seed driver (12 cells). Model on `run_issue270_ppo_continuation.sh`.
- `experiments/p3_specialization/analyze_282.py` (new) — Aggregate metrics across λ, compute `gap_closed_mean` per cell, emit tier verdict table. Model on `analyze_270.py` (gap_closed helper, reference table).
- `research_notebook/2026-05-17_issue282_lambda_verdict.md` (new, written after runs complete) — Verdict + tier classification + interpretation paragraph.
- (Reference only, not modified) `bucket_brigade/training/joint_trainer.py:202` — Trainer already accepts `gae_lambda`; no library-side change required.
- (Reference only, not modified) `bucket_brigade/training/networks.py:274,318` — `compute_gae` already parameterized correctly.

## Test Plan

- [ ] Manual verification (wiring): run `python experiments/p3_specialization/train.py --scenario minimal_specialization --lambda-red 0.0 --seed 42 --gae-lambda 0.99 --num-iterations 2 --rollout-steps 256 --output-dir /tmp/lambda_test --output-dir /tmp/lambda_smoke` (2 iters, tiny rollout) and confirm `config.json` records `gae_lambda: 0.99` and training completes without error. This is a wiring smoke test; do this locally (cheap, ~30s).
- [ ] Remote sweep: run `bash experiments/p3_specialization/run_issue282_lambda_sweep.sh` on `COMPUTE_HOST_PRIMARY` in tmux. Expected runtime ~30-60 min wall-clock for 12 cells at 2 seeds in parallel.
- [ ] Analyzer: run `python experiments/p3_specialization/analyze_282.py` against the sweep outputs and confirm it produces a per-λ table + tier verdict.
- [ ] Verdict writeup: produce `research_notebook/2026-05-17_issue282_lambda_verdict.md` with the tier classification, gap_closed per λ, and interpretation per the table above.
- [ ] Cross-check: λ=0.95 cell's `gap_closed_mean` should reproduce post-#236 baseline (~0.18 ± seed noise). Material drift here implies stack regression and blocks the verdict.
- [ ] Edge cases: confirm λ=1.0 cell completes without numerical issues (no NaN advantages — λ=1 means pure Monte Carlo return minus baseline, sensitive to value-function init; if it diverges, this is itself a verdict and should be documented).

### PR DESCRIPTION
**Pull request by rjwalters**
## Summary

Adds `--gae-lambda` CLI plumbing to `experiments/p3_specialization/train.py` and the issue #282 sweep + analyzer scaffolding. `JointPPOTrainer` already accepts `gae_lambda` as a constructor arg (`bucket_brigade/training/joint_trainer.py:202`) — this PR closes the wiring gap so the issue #282 4-λ × 3-seed sweep can be launched on a remote host.

The actual sweep (3 seeds × 4 λ values × 50 iters) is **remote-only** per `CLAUDE.md` compute guidelines and is **not** executed by this PR. Only local CLI plumbing and smoke tests were run.

## Changes

- `experiments/p3_specialization/train.py`:
  - Add `gae_lambda: float = 0.95` to `CellConfig`.
  - Add `--gae-lambda` argparse arg (default matches `CellConfig` / `JointPPOTrainer`).
  - Thread `gae_lambda=cfg.gae_lambda` into the `JointPPOTrainer(...)` constructor.
  - Log resolved λ in the cell-banner print and via `config.json` (existing `asdict(cfg)` dump).
- `experiments/p3_specialization/run_issue282_lambda_sweep.sh` (new): 4 λ ∈ {0.95, 0.99, 0.999, 1.0} × 3 seeds = 12 cells; mirrors `run_issue270_ppo_continuation.sh` (50 iters × 2048 steps, `--lambda-red 0.0`, IPPO).
- `experiments/p3_specialization/analyze_282.py` (new): per-λ trailing-5 `gap_closed` aggregation + tier verdict (≥0.5 / 0.25–0.5 / <0.25) + λ=0.95 cross-check against the PR #257 post-#236 baseline (~0.182). Models on `analyze_270.py`.

## Acceptance Criteria Verification

| Criterion | Status | Verification |
|-----------|--------|--------------|
| `train.py` exposes `--gae-lambda` | done | `train.py --help` lists the new flag with default 0.95. |
| `config.json` records resolved λ | done | Smoke run with `--gae-lambda 0.99` writes `"gae_lambda": 0.99` to `/tmp/lambda_smoke_099/config.json`. |
| λ=0.95 (default) preserves pre-#282 behavior | done | Smoke run with `--gae-lambda 0.95` produces iter-0 team_reward = -62.875, identical to the iter-0 reward at λ=0.99 / λ=1.0 (env+policy init are deterministic before any update). |
| λ=1.0 (pure MC) runs without NaN | done | Smoke run with `--gae-lambda 1.0 --num-iterations 2 --rollout-steps 256` completes cleanly, iter 1 team_reward = -88.863. |
| Sweep driver covers 4 λ × 3 seeds | done | `run_issue282_lambda_sweep.sh` iterates `LAMBDAS=(0.95 0.99 0.999 1.0)` × `seeds={42,43,44}`; same `2 -P` parallelism as #270. |
| Analyzer computes `gap_closed` per cell with tier verdict | done | `analyze_282.py` aggregates per-λ trailing-5 `gap_closed`, picks best λ, classifies into `tier_1_breaks_plateau` / `tier_2_partial` / `tier_3_insufficient` / `no_data`, emits both JSON + markdown verdict. Reuses the canonical `gap_closed` helper + reference table (MINSPEC_RANDOM=-96.07, MINSPEC_SPECIALIST=-22.07). |
| Full sweep NOT executed | done | Per user constraint and `CLAUDE.md`: 3 seeds × 4 λ × 50 iters is remote-only. This PR ships infrastructure only. |

## Test Plan

Local smoke (all ran, all passed):

- [x] `train.py --help` exposes `--gae-lambda`.
- [x] `train.py --gae-lambda 0.95 --num-iterations 2 --rollout-steps 256` runs; `config.json` records `gae_lambda: 0.95`.
- [x] `train.py --gae-lambda 0.99 --num-iterations 2 --rollout-steps 256` runs; `config.json` records `gae_lambda: 0.99`.
- [x] `train.py --gae-lambda 1.0 --num-iterations 2 --rollout-steps 256` runs without NaN (numerical stability check called out in the curator spec).
- [x] `analyze_282.py` parses CLI; no-data path produces `tier: no_data` cleanly.

Remote (out of scope for this PR — to be executed after merge):

- [ ] `bash experiments/p3_specialization/run_issue282_lambda_sweep.sh` on `COMPUTE_HOST_PRIMARY` in tmux (~30-60 min).
- [ ] `python experiments/p3_specialization/analyze_282.py` against the sweep output.
- [ ] λ=0.95 cell reproduces post-#236 baseline (~0.182 ± 0.10); the analyzer's `baseline_cross_check` block will flag a stack regression if not.

Closes #282

### CODE REVIEW

**Comment by rjwalters** [comment]
LGTM: foundational CLI plumbing for issue #282 (high-λ GAE sweep). Approved.

## Code Review

- `train.py`: `--gae-lambda` argparse arg added with `default = CellConfig.__dataclass_fields__["gae_lambda"].default` (DRY between dataclass + CLI). Threaded into `JointPPOTrainer(gae_lambda=cfg.gae_lambda)` (line 656). Logged in cell banner and recorded in `config.json` via the existing `asdict(cfg)` dump.
- `CellConfig.gae_lambda: float = 0.95` matches `JointPPOTrainer.__init__` default at `bucket_brigade/training/joint_trainer.py:202`. **Default-preservation property holds** — bit-identical pre-#282 behavior when flag is omitted.
- `analyze_282.py`: tier thresholds (`TIER_1_THRESHOLD = 0.50`, `TIER_2_THRESHOLD = 0.25`) match the curator spec. Reuses canonical `gap_closed` helper + MINSPEC reference table. Baseline cross-check vs PR #257 (~0.182) is solid guardrail against stack regression. `_lambda_tag` (`str(lam).replace(".", "_")`) matches sweep driver's `${lam//./_}` path convention — paths align between writer/reader.
- `run_issue282_lambda_sweep.sh`: 4 λ × 3 seeds = 12 cells, `-P 2`, 50 iters × 2048 steps. Mirrors `run_issue270_ppo_continuation.sh`. Remote-only per CLAUDE.md — not executed by this PR (intentional).

## Fixed During Evaluation

- `analyze_282.py` failed `ruff format` (one set-comprehension that needed line-collapsing). Applied `ruff format` directly and pushed to `feature/issue-282`. All 6 CI checks now pass: lint-python, pre-commit, test-python, test-web, build-web, integration-test.

## Stacking Note

PR #304 (issue #285, BC-init + high-λ) is stacked on this. Foundation verified clean — bugs here would have cascaded but did not.
